h2o-algos/src/main/java/hex/svm/IncompleteCholeskyFactorization.java

```diff
@@ -0,0 +1,278 @@
+      }
+    } finally {
+      workspace.delete();
+    }
+    return icf;
+  }
+
+  /**
+   * Calculate a new column of the ICF matrix
+   */
+  private static class CalculateColumn extends MRTask<CalculateColumn> {
+    // IN
+    Frame _full_frame;
+    DataInfo _dinfo;
+    Kernel _kernel;
+    double[] _header_row;
+    DataInfo.Row _pivot_sample;
+    Frame _icf;
+
+    private CalculateColumn(Frame frame, DataInfo dinfo, Kernel kernel, 
+                            Frame icf,
+                            DataInfo.Row pivotSample, double[] headerRow) {
+      _full_frame = frame;
+      _dinfo = dinfo;
+      _kernel = kernel;
+      _icf = icf;
+      _pivot_sample = pivotSample;
+      _header_row = headerRow;
+    }
+
+    @Override
+    public void map(Chunk pivot_selected, Chunk diag2, Chunk newColChunk) {
+      Chunk[] icf = getLocalChunks(_icf, pivot_selected.start());
+      Chunk[] frameChunks = getLocalChunks(_full_frame, pivot_selected.start());
+
+      double[] newColData = MemoryManager.malloc8d(newColChunk._len);
+      boolean[] pivotSelected = MemoryManager.mallocZ(newColChunk._len); 
+      for (int i = 0; i < newColData.length; i++) {
+        pivotSelected[i] = pivot_selected.at8(i) != 0;
+        newColData[i] = pivotSelected[i] ? newColChunk.atd(i) : 0;
+      }
+
+      for (int k = 0; k < icf.length - 1; k++) {
+        for (int i = 0; i < newColChunk._len; i++) {
+          if (pivotSelected[i])
+            continue;
+          newColData[i] -= icf[k].atd(i) * _header_row[k];
+        }
+      }
+      DataInfo.Row row = _dinfo.newDenseRow();
+      for (int i = 0; i < newColChunk._len; i++) {
+        if (pivotSelected[i])
+          continue;
+        _dinfo.extractDenseRow(frameChunks, i, row);
+        newColData[i] += _kernel.calcKernelWithLabel(row, _pivot_sample);
+      }
+      for (int i = 0; i < newColChunk._len; i++) {
+        if (pivotSelected[i])
+          continue;
+        newColData[i] /= _header_row[_header_row.length - 1];
+      }
+
+      // Updated newCol chunk and calculate diag2
+      for (int i = 0; i < newColData.length; i++) {
+        double v = newColData[i];
+        newColChunk.set(i, v);
+        diag2.set(i, diag2.atd(i) + (v * v));
+      }
+    }
+  }
+  
+  private static class InitICF extends MRTask<InitICF> {
+    // IN
+    DataInfo _dinfo;
+    Kernel _kernel;
+
+    InitICF(DataInfo dinfo, Kernel kernel) {
+      _dinfo = dinfo;
+      _kernel = kernel;
+    }
+
+    @Override
+    public void map(Chunk[] cs, NewChunk nc) {
+      DataInfo.Row row = _dinfo.newDenseRow();
+      Chunk two_norm_sq = cs[cs.length - 1];
+
+      for (int r = 0; r < cs[0]._len; r++) {
+        _dinfo.extractDenseRow(cs, r, row);
+        final double tns = row.twoNormSq();
+        row.response[1] = tns;
+        two_norm_sq.set(r, tns);
+        double diag1val = _kernel.calcKernel(row, row);
+        nc.addNum(diag1val);
+      }
```

Comment:
Calculate and set the diagonal values of Q

Reply:
It looks like the content of diag1 is set once and never changed.  However, during each iteration, the rows that are not in the set of pivots are changed during each iteration until either we hit the rank or that row is chosen as the one with the highest value.  Not sure this is accurate....

Reply:
Good observation Wendy!

Vector `v` (using notation of the paper) is made of two components: `diag1` and `diag2`. The implementation only updates `diag2`.

When we are finding the pivot we have take that into account: https://github.com/h2oai/h2o-3/pull/3448/files#diff-da32ad2a741144bea660b1709b30abacR187